### PR TITLE
Client Map EntryProcessor implementation

### LINKED ISSUE #92: All hazelcast exception should support builtin .Net serialization.
Exception like HazelcastException can be marked as Serializable to support containers some remote calls etc.

### PR DESCRIPTION
**Pull request by asimarslan**
* Client Map EntryProcessor implementation
* All exceptions marked as Serializable (fix #92 )

### CODE REVIEW

File: Hazelcast.Net/Hazelcast.Map/IEntryProcessor.cs [hunk 1]
```diff
@@ -12,27 +12,17 @@
...
 namespace Hazelcast.Map
 {
-    public interface IEntryProcessor<TKey, TValue>
+    /// <summary>
+    /// A stateful serializable object which represents the EntryProcessor defined on server side.
+    /// </summary>
+    /// <remarks>
+    /// This object must have a hazelcast serializable EntryProcessor counter part registered on server side with the actual
+    /// <c>org.hazelcast.map.EntryProcessor</c> implementation.
+    /// Client side EntryProcessor does not have any process logic.
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Map/IEntryProcessor.cs]: typo, "processing" logic

File: Hazelcast.Net/Hazelcast.Map/IEntryProcessor.cs [hunk 2]
```diff
@@ -12,27 +12,17 @@
...
-using System.Collections.Generic;
-
 namespace Hazelcast.Map
 {
-    public interface IEntryProcessor<TKey, TValue>
+    /// <summary>
+    /// A stateful serializable object which represents the EntryProcessor defined on server side.
+    /// </summary>
+    /// <remarks>
+    /// This object must have a hazelcast serializable EntryProcessor counter part registered on server side with the actual
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Map/IEntryProcessor.cs]: typo "counterpart"

File: Hazelcast.Net/Hazelcast.Core/IMap.cs [hunk 3]
```diff
@@ -294,6 +294,30 @@ public interface IMap<TKey, TValue> : IDistributedObject
...
+        object ExecuteOnKey(TKey key, IEntryProcessor entryProcessor);
+
+        /// <summary>Applies the user defined EntryProcessor to the entries mapped by the collection of keys.</summary>
+        /// <param name="keys">keys to process</param>
+        /// <param name="entryProcessor">the user defined <see cref="IEntryProcessor"/></param>
+        /// <returns>the results mapped by each key in the collection.</returns>
+        IDictionary<TKey, object> ExecuteOnKeys(ISet<TKey> keys, IEntryProcessor entryProcessor);
+
+        /// <summary>Applies the user defined EntryProcessor to the all entries in the map.</summary>
+        /// <param name="entryProcessor">the user defined the user defined <see cref="IEntryProcessor"/></param>
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Core/IMap.cs]: typo double "the user defined". I think these docs should also include the fact that the actual implementation must be on the server side.

File: Hazelcast.Net/Hazelcast.Core/IMap.cs [hunk 4]
```diff
@@ -294,6 +294,30 @@ public interface IMap<TKey, TValue> : IDistributedObject
         /// </remarks>
         void EvictAll();
 
+        /// <summary>Applies the user defined EntryProcessor to the entry mapped by the key.</summary>
+        /// <param name="key">key</param>
+        /// <param name="entryProcessor">the user defined the user defined <see cref="IEntryProcessor"/></param>
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Core/IMap.cs]: typo double "the user defined", and see other comment regarding docs

File: Hazelcast.Net/Hazelcast.Core/IMap.cs [hunk 5]
```diff
@@ -294,6 +294,30 @@ public interface IMap<TKey, TValue> : IDistributedObject
...
+        /// <param name="entryProcessor">the user defined <see cref="IEntryProcessor"/></param>
+        /// <returns>the results mapped by each key in the collection.</returns>
+        IDictionary<TKey, object> ExecuteOnKeys(ISet<TKey> keys, IEntryProcessor entryProcessor);
+
+        /// <summary>Applies the user defined EntryProcessor to the all entries in the map.</summary>
+        /// <param name="entryProcessor">the user defined the user defined <see cref="IEntryProcessor"/></param>
+        /// <returns></returns>
+        IDictionary<TKey, object> ExecuteOnEntries(IEntryProcessor entryProcessor);
+
+        /// <summary>Applies the user defined EntryProcessor to the all entries in the map.</summary>
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Core/IMap.cs]: doc is wrong, this only applies to entries matching a certain predicate

**Comment by asimarslan** [comment]
run-lab-run

**Comment by asimarslan** [comment]
verify

File: Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapProxy.cs [hunk 6]
```diff
@@ -453,6 +467,60 @@ public void EvictAll()
...
+                return new Dictionary<TKey, object>();
+            }
+            var dataList = ToDataList(keys);
+            var request = MapExecuteOnKeysCodec.EncodeRequest(GetName(), ToData(entryProcessor), dataList);
+            var response = Invoke(request);
+            var resultParameters = MapExecuteOnKeysCodec.DecodeResponse(response);
+            return PrepareResult(resultParameters.response);
+        }
+
+        internal IDictionary<TKey, object> PrepareResult(IList<KeyValuePair<IData, IData>> entries) {
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapProxy.cs]: can the method be private?

File: Hazelcast.Net/Hazelcast.Core/IMap.cs [hunk 7]
```diff
@@ -925,6 +967,29 @@ public interface IMap<TKey, TValue> : IDistributedObject
...
+//        /// <summary>
+//        /// Applies the user defined EntryProcessor to the entry mapped by the key with specified IExecutionCallback to
+//        /// listen event status and returns immediately.
+//        /// </summary>
+//        /// <param name="key">key to be processed</param>
+//        /// <param name="entryProcessor">the user defined <see cref="IEntryProcessor"/></param>
+//        /// <param name="executionCallback">to listen whether operation is finished or not</param>
+//        /// <typeparam name="TResp">Response Type</typeparam>
+//        void SubmitToKey<TResp>(TKey key, IEntryProcessor entryProcessor, IExecutionCallback<TResp> executionCallback);
+
```
Comment (cangencer) [inline Hazelcast.Net/Hazelcast.Core/IMap.cs]: commented out can be removed

File: Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapProxy.cs (hunk 6 above)
Comment (asimarslan) [inline Hazelcast.Net/Hazelcast.Client.Proxy/ClientMapProxy.cs]: I'll move it to ClientProxy

**Comment by asimarslan** [comment]
verify

**Comment by asimarslan** [comment]
verify

**Pull request merged**